feature: Compose Multiplatform for web, browser navigation support (#316)
JetBrains/kotlin-multiplatform-dev-docs: topics/compose/compose-navigation-routing.md

as well as use the address bar to understand where they are and get to a destination directly.

To bind the web app to the navigation graph defined in your common code,
call the `window.bindToNavigation()` method in your Kotlin/Wasm or Kotlin/JS code:

```kotlin
@OptIn(ExperimentalComposeUiApi::class)
@ExperimentalBrowserHistoryApi
fun main() {
    val body = document.body ?: return

    ComposeViewport(body) {

        val navController = rememberNavController()
        // Assuming that your main composable function in common code is App()
        App(navController)
        LaunchedEffect(Unit) {
            window.bindToNavigation(navController)
        }
    }
}
```

After a `window.bindToNavigation(navController)` call:
* The URL displayed in the browser reflects the current route (in the URL fragment, after the `#` character).
* The app parses URLs entered manually to translate them into destinations within the app.

By default, when using type-safe navigation, a destination is translated into a URL fragment using this pattern:
`<app package>.<type parameter of the destination>_<argument1>_<argument2>`.

> For example, `example.org#org.example.app.StartScreen_123_Alice%2520Smith`.

Reviewer: no, it depends on the serialization config.
```kotlin
@Serializable @SerialName("screen1") data object Screen1
```
will be shown as `#screen1`

Author: Got it, thanks. This section is about the default definition, which should exclude additional configuration, but I'll put this in "Customize" — along with the general recommendation according to the other comment to try `@SerialName` first, and `getBackStackEntryRoute` for surgical manipulations.

### Customize translating routes into URLs and back

> 

Reviewer: I would add here that usually @SerialName is enough for customisations. And `getBackStackEntryRoute` is need for rare manual cases

> As Compose Multiplatform apps are single-page apps, the framework needs to manipulate the address bar to imitate

Reviewer 2: ```suggestion
As Compose Multiplatform apps are single-page apps, the framework manipulates the address bar to imitate
```

usual web navigation.
If you would like to make your URLs more readable or isolate implementation from URL patterns,

> you can develop custom processing for destination routes in your Kotlin/Wasm or Kotlin/JS code:

Reviewer 2: ```suggestion
you can define custom processing for destination routes in your Kotlin/Wasm or Kotlin/JS code:
```

1. Pass the optional `getBackStackEntryRoute` lambda to the `window.bindToNavigation()` function

>     to describe converting routes into URL fragments where necessary.

Reviewer 2: ```suggestion
    to specify how routes should be converted into URL fragments when needed.
```

                        // Accesses the route arguments.
                        val args = entry.toRoute<Id>()

                        // Sets the corresponding URL fragment to "#find_id_222"
                        // instead of "#org.example.app.ID%2F222"
                        "#find_id_${args.id}"
                    }
                    route.startsWith(Patient.serializer().descriptor.serialName) -> {
                        val args = entry.toRoute<Patient>()
                        // Sets the corresponding URL fragment to "#patient_Jane%20Smith-Baker_33"
                        // instead of "#org.company.app.Patient%2FJane%2520Smith-Baker%2F33"
                        "#patient_${args.name}_${args.age}"
                    }
                    // Doesn't set a URL fragment for all other routes.
                    else -> ""
                }
            }
        }
    }
}
```
<!--{default-state="collapsed" collapsible="true" collapsed-title="window.bindToNavigation(navController) { entry ->"}-->

> Make sure that every string that corresponds to a route starts with the `#` character to keep the data
> within URL fragments.

> > Otherwise, when the user copies and pastes the URL, the browser will try to access a wrong endpoint instead of passing

Reviewer 2: ```suggestion
> Otherwise, when users copy and paste the URL, the browser will try to access a wrong endpoint instead of passing
```

> the control to your app.
> 
{style="note"}

When your URLs have custom formatting, you should add the reverse processing 
to match a manually entered URL to a destination route.
The code that does the matching needs to run before the `window.bindToNavigation()` call binds
`window.location` to the navigation graph:

```kotlin
@OptIn(
    ExperimentalComposeUiApi::class,
    ExperimentalBrowserHistoryApi::class,
    ExperimentalSerializationApi::class
)
fun main() {
    val body = document.body ?: return
    ComposeViewport(body) {
        val navController = rememberNavController()
        App(navController)
        LaunchedEffect(Unit) {
            // Accesses the fragment substring of the current URL.
            val initRoute = window.location.hash.substringAfter('#', "")
            when  {

>                 // Identifies the corresponding route and navigates to it.

Reviewer 2: ```suggestion
                // Identifies the corresponding route and navigates to it
```

                initRoute.startsWith("start") -> {
                    navController.navigate(StartScreen)
                }
                initRoute.startsWith("find_id") -> {

>                     // Parses the string to extract route parameters before navigating to it.

Reviewer 2: ```suggestion
                    // Parses the string to extract route parameters before navigating to it
```

        LaunchedEffect(Unit) {
            window.bindToNavigation(navController)
        }
    }
}
```

After a `window.bindToNavigation(navController)` call:
* The URL displayed in the browser reflects the current route (in the URL fragment, after the `#` character).
* The app parses URLs entered manually to translate them into destinations within the app.

By default, when using type-safe navigation, a destination is translated into a URL fragment using this pattern:
`<app package>.<type parameter of the destination>_<argument1>_<argument2>`.
For example, `example.org#org.example.app.StartScreen_123_Alice%2520Smith`.

### Customize translating routes into URLs and back

As Compose Multiplatform apps are single-page apps, the framework manipulates the address bar to imitate
usual web navigation.
If you would like to make your URLs more readable and isolate implementation from URL patterns,
you can assign the name to a screen directly or develop fully custom processing for destination routes:

* To simply make a URL readable, use the `@SerialName` annotation for a serializable object or class:

    ```kotlin
    // Instead of using the app package and object name,

>     // this route will be translated to the URL simply as "#start".

Reviewer 2: ```suggestion
    // this route will be translated to the URL simply as "#start"
```
